packages/hdwallet-core/src/bitcoin.ts

```diff
@@ -37,63 +47,179 @@ export interface BitcoinOutput {
 
 /**
  * Output for a transaction we're about to sign.
  */
-export interface BTCSignTxOutput {
+export type BTCSignTxOutputSpend = {
+  addressType?: BTCOutputAddressType.Spend;
+  amount: string;
+  address: string;
+};
+
+// Will make address type more specific on TS version bump.
+export type BTCSignTxOutputSpendP2PKH = {
+  addressType?: BTCOutputAddressType.Spend;
+  amount: string;
+  address: string;
+  scriptType: BTCOutputScriptType.PayToAddress;
+};
+
+export type BTCSignTxOutputSpendP2SH = {
+  addressType?: BTCOutputAddressType.Spend;
+  amount: string;
+  address: string;
+  scriptType: BTCOutputScriptType.PayToMultisig | BTCOutputScriptType.PayToP2SHWitness;
+};
+
+export type BTCSignTxOutputSpendP2WPKH = {
+  addressType?: BTCOutputAddressType.Spend;
+  amount: string;
+  address: string;
+  scriptType: BTCOutputScriptType.PayToWitness;
+};
+
+export type BTCSignTxOutputTransfer = {
+  addressType: BTCOutputAddressType.Transfer;
+  amount: string;
   /** bip32 path for destination (device must `btcSupportsSecureTransfer()`) */
-  addressNList?: BIP32Path;
-  scriptType?: BTCOutputScriptType;
-  address?: string;
-  addressType: BTCOutputAddressType;
+  addressNList: BIP32Path;
+  scriptType: BTCOutputScriptType;
+};
+
+export type BTCSignTxOutputChange = {
+  addressType: BTCOutputAddressType.Change;
   amount: string;
-  isChange: boolean;
+  /** bip32 path for destination (device must `btcSupportsSecureTransfer()`) */
+  addressNList: BIP32Path;
+  scriptType: BTCOutputScriptType;
+  isChange: true;
+};
+
+export type BTCSignTxOutputExchange = {
   /**
    * Device must `btcSupportsNativeShapeShift()`
    */
-  exchangeType?: ExchangeType;
-  opReturnData?: string;
-}
+  addressType: BTCOutputAddressType.Exchange;
+  amount: string;
+  exchangeType: ExchangeType;
+};
+
+export type BTCSignTxOutputMemo = {
+  addressType?: BTCOutputAddressType.Spend;
+  amount?: "0";
+  opReturnData: string | Uint8Array;
+};
+
+export type BTCSignTxOutput = GuardedUnion<
+  | BTCSignTxOutputSpend
+  | BTCSignTxOutputSpendP2PKH
+  | BTCSignTxOutputSpendP2SH
+  | BTCSignTxOutputSpendP2WPKH
+  | BTCSignTxOutputTransfer
+  | BTCSignTxOutputChange
+  | BTCSignTxOutputExchange
+  | BTCSignTxOutputMemo
+>;
 
 export interface BTCSignTx {
   coin: string;
   inputs: Array<BTCSignTxInput>;
   outputs: Array<BTCSignTxOutput>;
   version?: number;
   locktime?: number;
-  opReturnData?: string;
+  opReturnData?: string; // TODO: dump this in favor of BTCSignTxOutputMemo above
   vaultAddress?: string;
 }
 
+export type BTCSignTxKK = Omit<BTCSignTx, "inputs"> & { inputs: Array<BTCSignTxInputKK> };
+export type BTCSignTxNative = Omit<BTCSignTx, "inputs"> & { inputs: Array<BTCSignTxInputNative> };
+export type BTCSignTxTrezor = Omit<BTCSignTx, "inputs"> & { inputs: Array<BTCSignTxInputTrezor> };
+export type BTCSignTxLedger = Omit<BTCSignTx, "inputs"> & { inputs: Array<BTCSignTxInputLedger> };
```

Comment:
as discussed above, normalize BTCSignTxInput and we can dump these special types.